### PR TITLE
Homework 4

### PR DESCRIPTION
**Pull request by peironggg**
Click [here](https://github.com/peironggg/CIS194/tree/04-higher-order/04-higher-order/homework-04) for source code and question pdf.

Thank you for your feedback 😄

### CODE REVIEW

File: 04-higher-order/homework-04/hw-04.hs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+-}
+--Exercise 3
+xor :: [Bool] -> Bool
+xor = foldr (/=) False
+
+map' :: (a -> b) -> [a] -> [b]
+map' f = foldr (\x xs -> f x : xs) []
+
+myFoldl :: (a -> b -> a) -> a -> [b] -> a
+myFoldl f initial xs = foldr (flip f) initial (reverse xs)
```
Comment (jhrcek) [inline 04-higher-order/homework-04/hw-04.hs:140]: This implementation works fine, but there's a way to do it without relying on any existing list functions (apart from foldr of course :) )
This is rather hard, but can be done if you realize you need to accumulate a function and in the end you apply it to the initial element. Only try this is you want extra challenge. If you give you google for "implementing foldl in terms of foldl".

**Review by jhrcek — APPROVED** [review]
There are no clear improvements I can think of, nice work :+1:

File: 04-higher-order/homework-04/hw-04.hs (hunk 1 above)
Comment (peironggg) [reply 04-higher-order/homework-04/hw-04.hs:140]: @jhrcek Thanks for the feedback, now I know at which direciton to look into. I wasn't very satisfied with my answer anyways 😂

File: 04-higher-order/homework-04/hw-04.hs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+myFoldl f base xs = foldr ...
+in such a way that myFoldl behaves identically to the standard
+foldl function.
+Hint: Study how the application of foldr and foldl work out:
+foldr f z [x1, x2, ..., xn] == x1 ‘f‘ (x2 ‘f‘ ... (xn ‘f‘ z)...)
+foldl f z [x1, x2, ..., xn] == (...((z ‘f‘ x1) ‘f‘ x2) ‘f‘...) ‘f‘ xn
+-}
+--Exercise 3
+xor :: [Bool] -> Bool
+xor = foldr (/=) False
```
Comment (kwannoel) [inline 04-higher-order/homework-04/hw-04.hs:134]:
```suggestion
-- import Data.Foldable (foldl')
xor = foldl' (/=) False
```

If we run benchmarks:

```
benchmarking all true/foldr
time                 95.74 ms   (85.65 ms .. 101.8 ms)
                     0.989 R²   (0.970 R² .. 1.000 R²)
mean                 91.96 ms   (80.80 ms .. 100.4 ms)
std dev              15.69 ms   (9.977 ms .. 26.64 ms)
variance introduced by outliers: 54% (severely inflated)

benchmarking all true/foldl'
time                 7.265 ms   (7.158 ms .. 7.366 ms)
                     0.999 R²   (0.998 R² .. 0.999 R²)
mean                 7.183 ms   (7.126 ms .. 7.249 ms)
std dev              171.7 μs   (150.0 μs .. 204.9 μs)

benchmarking all false/foldr
time                 22.12 ms   (21.67 ms .. 22.57 ms)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 22.69 ms   (22.48 ms .. 22.92 ms)
std dev              497.0 μs   (378.5 μs .. 731.1 μs)

benchmarking all false/foldl'
time                 6.612 ms   (6.491 ms .. 6.771 ms)
                     0.997 R²   (0.995 R² .. 0.999 R²)
mean                 6.520 ms   (6.466 ms .. 6.602 ms)
std dev              195.0 μs   (143.0 μs .. 247.2 μs)
variance introduced by outliers: 13% (moderately inflated)

benchmarking alternating/foldr
time                 65.99 ms   (58.06 ms .. 71.84 ms)
                     0.968 R²   (0.913 R² .. 0.996 R²)
mean                 63.10 ms   (56.35 ms .. 68.85 ms)
std dev              11.45 ms   (8.236 ms .. 15.77 ms)
variance introduced by outliers: 62% (severely inflated)

benchmarking alternating/foldl'
time                 7.647 ms   (7.471 ms .. 7.835 ms)
                     0.995 R²   (0.991 R² .. 0.998 R²)
mean                 7.501 ms   (7.393 ms .. 7.632 ms)
std dev              336.0 μs   (261.2 μs .. 432.2 μs)
variance introduced by outliers: 22% (moderately inflated)
```

It appears that using `foldl'` is significantly faster than `foldr`. Why do you think this is so?

[benchmark script](https://gist.github.com/kwannoel/567d7795c67907ce1bb9a4cbc086e8bc).

Comment (kwannoel) [reply 04-higher-order/homework-04/hw-04.hs:134]: Once you figure out, you will see other areas in your code in which using `foldl'` will improve performance.

Comment (peironggg) [reply 04-higher-order/homework-04/hw-04.hs:134]: @kwannoel Thanks for explaining the differences in speed using the benchmark tests.

**Pull request merged**